interpreter/luajit/luajit.go

```diff
@@ -0,0 +1,450 @@
+	base := path.Base(info.FileName())
+	if !strings.HasPrefix(base, "libluajit-5.1.so") && base != "luajit" {
+		return nil, nil
+	}
+
+	ef, err := info.GetELF()
+	if err != nil {
+		return nil, err
+	}
+
+	luaInterp, err := extractInterpreterBounds(info.Deltas(), cframeSize)
+	if err != nil {
+		return nil, err
+	}
+	logf("lj: interp range %v", luaInterp)
+
+	ljd := &luajitData{}
+
+	if err = extractOffsets(ef, ljd); err != nil {
+		return nil, err
+	}
+
+	logf("lj: offsets %+v", ljd)
+
+	if err = ebpf.UpdateInterpreterOffsets(support.ProgUnwindLuaJIT, info.FileID(),
+		[]util.Range{luaInterp}); err != nil {
+		return nil, err
+	}
+
+	return ljd, nil
+}
+
+// LuaJIT's interpreter isn't a function, its a raw chunk of assembly code with direct threaded
+// jumps at end of each opcode. The public entrypoints (lua_pcall/lua_resume) call the lj_vm_pcall
+// function at the end of this blob which set up the interpreter and starts executing.
+// Even though its not a normal function an eh_frame entry is created for it, its really
+// big and has a somewhat unique FDE we can pick out. We could tighten this up by looking for
+// direct jumps to the start of the interpreter (one can be found lj_dispatch_update) but we'd
+// still need to consult the stack deltas to get the end of the interpreter.
+func extractInterpreterBounds(deltas stackdeltatypes.StackDeltaArray, param int32) (util.Range,
+	error) {
+	for i := 0; i < len(deltas)-1; i++ {
+		d, next := &deltas[i], &deltas[i+1]
+		if next.Address-d.Address > 15000 {
+			// The case covers x86 w/ dwarf and old versions of luajit that used dwarf and
+			// the second covers more recent arm versions that use frame pointers.
+			if d.Info.Opcode == stackdeltatypes.UnwindOpcodeBaseSP && d.Info.Param == param ||
+				d.Info.Opcode == stackdeltatypes.UnwindOpcodeBaseFP && d.Info.Param == 16 {
+				return util.Range{Start: d.Address, End: next.Address}, nil
+			}
+		}
+	}
+
+	return util.Range{}, errors.New("failed to find interpreter range")
+}
+
+func (l *luajitInstance) getVMList() []libpf.Address {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	gs := make([]libpf.Address, 0, len(l.vms))
+	for g := range l.vms {
+		gs = append(gs, g)
+	}
+	return gs
+}
+
+func (l *luajitInstance) addJITRegion(ebpf interpreter.EbpfHandler, fim process.FileIDMapper, pr process.Process,
+	start, end uint64) error {
+	prefixes, err := lpm.CalculatePrefixList(start, end)
+	if err != nil {
+		logf("lj: failed to calculate lpm: %v", err)
+		return err
+	}
+	logf("lj: add JIT region pid(%v) %#x:%#x", pr.PID(), start, end)
+	for _, prefix := range prefixes {
+		// TODO: fix these: WARN[0267] Failed to lookup file ID 0x2a00000000
+		fileID := uint64(C.LUAJIT_JIT_FILE_ID) << 32
+		fim.Set(host.FileID(fileID), libpf.NewFileID(fileID, 0))
+		if err := ebpf.UpdatePidInterpreterMapping(pr.PID(), prefix, support.ProgUnwindLuaJIT,
+			host.FileID(fileID), 0); err != nil {
+			return err
+		}
+		l.prefixes[start] = prefix
```

Comment:
Nice catch, this is broken in theory but works in practice because for huge VM ranges (luajit's smallest is 64k) prefixes will always have one entry.  But that's crap and I should it fix it to be general.  One of those "I'll go back and fix this later" short cuts I neglected to revisit.